File: 2018/day-2/node/main.js
var fs = require("fs");

console.log("-----------------------------");
console.log("Advent of Code - Day 2 - Node");

var contents = fs.readFileSync("../input.txt", "utf8");
var ids = contents.split("\n");

function checksum(ids) {
  var twos = 0;
  var threes = 0;

  for (let i = 0; i < ids.length; i++) {
    const id = ids[i];
    charactersInID = {};
    twoIdenticalFound = false;
    threeIdenticalFound = false;

    for (let j = 0; j < id.length; j++) {
      charactersInID[id[j]] = (charactersInID[id[j]] || 0) + 1;
    }

    // Find characters that are either double or tripple
    for (const key of Object.keys(charactersInID)) {
      if (!twoIdenticalFound && charactersInID[key] == 2) {
        twos++;
        twoIdenticalFound = true; // Increment only once per ID
      }

      if (!threeIdenticalFound && charactersInID[key] == 3) {
        threes++;
        threeIdenticalFound = true; // Increment only once per ID
      }
    }
  }
  //   console.log(
  //     "Words with two characters: ",
  //     twos,
  //     "and three characters: ",
  //     threes
  //   );
  return twos * threes;
}

function commonLettersOfCorrectBoxes(ids) {
  for (let i = 0; i < ids.length; i++) {
    const id1 = ids[i];
    for (let j = i + 1; j < ids.length; j++) {
      const id2 = ids[j];
      let identical = "";

      // Compare all characters of string
      for (let n = 0; n < id1.length; n++) {
        if (id1[n] == id2[n]) {
          identical += id1[n];
        }
      }

      // Check if the IDs only differ by 1 character
      if (identical.length == id1.length - 1) {
        return identical;
      }
    }
  }
  return "";
}

console.log("Part 1");
console.log("- Checksum of IDs is:", checksum(ids));
console.log("Part 2");
console.log(
  "- Common characters of correct boxes:",
  commonLettersOfCorrectBoxes(ids)
);

File: 2018/day-2/go/main.go
package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"strings"
)

func hasTwoThreeIdentical(s string) (bool, bool) {
	charMap := make(map[rune]int)
	for _, c := range s {
		charMap[c]++
	}
	hasTwo := false
	hasThree := false
	for _, v := range charMap {
		switch v {
		case 2:
			hasTwo = true
		case 3:
			hasThree = true
		default:
		}
	}
	return hasTwo, hasThree
}

func printCheckSum(filename string) error {
	f, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer f.Close()

	var countTwo, countThree int
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		hasTwo, hasThree := hasTwoThreeIdentical(scanner.Text())

		if hasTwo {
			countTwo++
		}
		if hasThree {
			countThree++
		}
	}

	checkSum := countTwo * countThree
	fmt.Printf("- The checksum for boxes with twos (%d) and threes (%d) is: %d\n", countTwo, countThree, checkSum)
	return nil
}

func findAlmostEquals(filename string) error {
	bytes, err := ioutil.ReadFile(filename)
	if err != nil {
		return err
	}
	ids := strings.Split(string(bytes), "\n")

	// Compare all IDs to all other IDs
	for i := 0; i < len(ids); i++ {
		id1 := ids[i]

		for j := i + 1; j < len(ids); j++ {
			id2 := ids[j]

			//Compare ID 1 to ID 2
			differencesFound := 0
			for k := 0; k < len(id1); k++ {

				if id1[k] != id2[k] {
					differencesFound++
				}
			}

			if differencesFound == 1 {
				fmt.Print("- First almost right id: ")
				printIdenticalParts(id1, id2)
				return nil
			}
		}
	}
	return nil
}

func printIdenticalParts(id1, id2 string) {
	for k := 0; k < len(id1); k++ {
		if id1[k] == id2[k] {
			fmt.Print(string(id1[k]))
		}
	}
}

func main() {
	fmt.Println("Advent of Code - Day 2 - Go")

	// Print checksum for part 1
	fmt.Println("Part 1")
	err := printCheckSum("input.txt")
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("Part 2")
	err = findAlmostEquals("input.txt")
	if err != nil {
		log.Fatal(err)
	}

}
